### DISCUSSION TITLE
Incorrect polygon intersects results?

### DISCUSSION THREAD

**Discussion by sreekanth-cb** [google/s2geometry#251] [Q&A] [answered]
Hi @jmr,

I am getting incorrect polygon intersects results for the below test. (polygons aren't intersecting in real).
Not sure whether anything wrong with the input/test.

```

TEST(S2NonOverlapping1, AllCases) {
  vector<S2Point> pts(7);
  S2LatLng latlng;

  pts[0] = latlng.FromDegrees(12.980680426949517, 77.59652137756348).ToPoint();
  pts[1] = latlng.FromDegrees(12.976833078657933, 77.59218692779541).ToPoint();
  pts[2] = latlng.FromDegrees(12.978422207996038, 77.59034156799316).ToPoint();
  pts[3] = latlng.FromDegrees(12.969890973519856, 77.58613586425781).ToPoint();
  pts[4] = latlng.FromDegrees(12.964746999572585, 77.59210109710693).ToPoint();
  pts[5] = latlng.FromDegrees(12.972902030870115, 77.59969711303711).ToPoint();
  pts[6] = latlng.FromDegrees(12.979258587778183, 77.60179996490479).ToPoint();
  
  vector<std::unique_ptr<S2Loop>> loops;
  S2Loop* loop = new S2Loop(pts);
  EXPECT_EQ(true, loop->IsValid());

  loops.emplace_back(loop);  
  S2Polygon polygon1 = S2Polygon(std::move(loops));
  EXPECT_EQ(true,polygon1.IsValid()); 

  pts.clear();
  pts.resize(3);
  pts[0] = latlng.FromDegrees(63.03333333333333, 90.13333333333334).ToPoint();
  pts[1] = latlng.FromDegrees(63.06666666666667, 90).ToPoint();
  pts[2] = latlng.FromDegrees(63.3, 90.13333333333334).ToPoint();
 
  loop = new S2Loop(pts);
  EXPECT_EQ(true, loop->IsValid());

  loops.clear();
  loops.emplace_back(loop);  
  auto polygon2 = S2Polygon(std::move(loops));
  EXPECT_EQ(true,polygon2.IsValid()); 

  bool intersects = polygon1.Intersects(&polygon2);
  EXPECT_EQ(false,intersects); 
}
```

Thanks!

**Comment by jmr** [comment, marked as answer]
Isn't this the same as #226?  `polygon2` is reversed and covers nearly the whole sphere. Interior is to the left.

**Reply by sreekanth-cb** [reply]
Thanks, @jmr for the quick response! 

Ah, Is there a way to figure out this orientation problem using this library itself?  (I was loading these points in online tools and hoping that their orientation is good enough)

My intent is to detect and flag this or auto-fix this issue before ingesting these polygon coordinates.

Something like,

if !loop.IsNormalized() {
 loop. Normalize() Or loop.Invert()

Or anything at the polygon level itself?

**Reply by sreekanth-cb** [reply]
```
if (!loop->IsNormalized()) {
    loop->Normalize();
  }
```
fixed the problem and I hope that is the idiomatic way of handling this with s2.

**Comment by jmr** [comment]
What's the source of the data? Don't they define the interior as being to the left? You can normalize like that, but then you can't have polygons that cover more than half the sphere, like a polygon for the oceans would be.

**Reply by sreekanth-cb** [reply]
Indeed, Sooner I had realized the pitfalls of normalizing like above. 
The data was just auto-generated by someone and have fixed the logic to create it in CCW manner and things are looking fine so far.

And the reason I didn't suspect the data was that we were comparing the results with ElasticSearch and things were working fine at the ES side, hence thought that everything was fine with the data.

Thanks!
